proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
+If `call_tag.canon` were not restricted, then this module could maliciously convert between `$abs` and `t` using the instructions `call_with_tag (call_tag.canon ([$abs] -> [t])) (func.ref $id_t) : [$abs] -> [t]` and `call_with_tag (call_tag.canon ([t] -> [$abs])) (func.ref $id_t) : [t] -> [$abs]`, where `(func $func_ref (param t) (result t) (return (local.get 0)))`.
+These instructions work because `call_tag.canon ([$abs] -> [t])` and `call_tag.canon ([t] -> [$abs])` are at run time both equal to `call_tag.canon ([t] -> [t])` when `$abs` happens to be instantiated with `t`.
 
-This has four applications:
-1. You can get more graceful behavior than a trap, e.g. throwing an exception.
-2. You can use this to support deferred loading, i.e. the "fall-back" function prompts the missing functionality to be loaded in.
-3. For dynamically typed object-oriented languages, you can use this to implement support user-specified handling of missing methods. That is, objects would have built-in methods, and the "fall-back" function would kick in whenever the method was not built into the object at creation time, in which case it can look for the method in the "added later" dictionary, and if that fails it can call the object's "missing method" method.
-4. For functional languages, this makes it possible to support *unbounded* dynamic arity. In particular, when you create a closure for a value of type `a -> b -> c` (where each letter is a type variable), the `funcref` in your closure can handle the unary and binary call tags. But if `c` is abstracting a function type, your `funcref` might get called with call tags for higher arity. Without a fall-back, you have to cap the arity so that this `funcref` can `func_switch` on a finite number of cases. With a fall-back, you can make the fall-back handler for an n-ary call tag dynamically look up the arity the closure was compiled with (in this example, `2`), call it with just that many arguments, and then call the returned closure with the remaining arguments (in this example, using the `n-2`-arity call tag).
+So the restriction on `call_tag.canon` respects type abstraction.
+Meanwhile, `call_tag.new` has no such restriction (because just canonicalization is what breaks abstraction) and so can be used to support dynamically typed function calls even in the presence of abstract types.
 
-### Variant
+### Simplification
 
-Many call tags are specialized/optimized cases of more generic call tags.
-In these cases, it can be useful to pass the `funcref` that didn't recognize the call tag to the fall-back handler so that a specialized call tag can simply call the same `funcref` with the more generic call tag.
+Call tags make it unnecessary for `funcref`s to have casting infrastructure, and they make the type-checking complexity of first-class function types unnecessary (which prior works in this space have found to be problematic).
```

Comment:
Not to ask for cherries on top, but it would be lovely to see this and the next section elaborated at some point, specifically the references to prior works here and the lack of support in typed function references below.